esp32/machine_pin.c

```diff
@@ -151,7 +151,11 @@ STATIC mp_obj_t machine_pin_obj_init_helper(const machine_pin_obj_t *self, size_
 
     // configure mode
     if (args[ARG_mode].u_obj != mp_const_none) {
-        gpio_set_direction(self->id, mp_obj_get_int(args[ARG_mode].u_obj));
+      if (self->id >= 34 && (mp_obj_get_int(args[ARG_mode].u_obj) & GPIO_MODE_DEF_OUTPUT)) {
+          nlr_raise(mp_obj_new_exception_msg(&mp_type_ValueError, "Pin can only be input"));
+        } else {
+          gpio_set_direction(self->id, mp_obj_get_int(args[ARG_mode].u_obj));
```

Comment:
Instead of calling `mp_obj_get_int(...)` here and above it would be more efficient (and less duplicated code) to create a variable (of type mp_int_t) to cache the return value of this function.